### ISSUE TITLE
Wrong result for CPU call to `triangular_solve` using a sparse CSR matrix LHS, a dense matrix RHS and `upper=False, transpose=True, unitriangular=True`

### ISSUE THREAD

**Issue by tvercaut** [pytorch/pytorch#88890]
### 🐛 Describe the bug

`triangular_solve` fails to provide a reasonable solution under the following conditions:
- CPU device
- Sparse CSR matrix LHS input (As per  #87358, the new `torch.linalg.solve_triangular` does not yet support sparse matrices)
- Dense matrix (i.e. more than a single vector) RHS input
- `upper=False, transpose=True, unitriangular=True` flags

Steps to reproduce:
```python
!pip3 install torch torchvision torchaudio torchtext --extra-index-url https://download.pytorch.org/whl/cu117 --upgrade

import torch
print(f'Running PyTorch version: {torch.__version__}')

torchdevices = [torch.device('cpu')]
if torch.cuda.is_available():
  torchdevices.append( torch.device('cuda') )
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))

for torchdevice in torchdevices:
  print('Running on ' + str(torchdevice))

  A_dense = torch.tensor([[0.0000, 0.0000, 0.0000, 0.0000],
          [2.6085, 0.0000, 0.0000, 0.0000],
          [0.0000, 0.1871, 0.0000, 0.0000],
          [0.0000, 0.9922, 0.0000, 0.0000]], device=torchdevice)
  A_csr = A_dense.to_sparse_csr()

  id_csr = torch.sparse_csr_tensor(torch.arange(4+1), torch.arange(4), torch.ones(4, device=torchdevice), (4,4), device=torchdevice)
  print("id_csr:\n",id_csr,"\n",id_csr.to_dense())

  b = torch.ones(4, 2, device=torchdevice)

  print("A_dense:\n",A_dense)
  print("b:\n",b)

  res = torch.triangular_solve(b, A_csr, upper=False, transpose=True, unitriangular=True).solution
  print("test1\n",res)

  res = torch.triangular_solve(b, A_csr.t().to_sparse_csr(), upper=True, unitriangular=True).solution
  print("test2\n",res)

  res = torch.triangular_solve(b, A_dense, upper=False, transpose=True, unitriangular=True).solution
  print("test3\n",res)

  res = torch.linalg.solve_triangular(A_dense.t(), b, upper=True, unitriangular=True)
  print("test4\n",res)

  res = torch.linalg.solve_triangular(A_dense, b.t(), upper=False, left=False, unitriangular=True).t()
  print("test5\n",res)

  res = torch.triangular_solve(b, A_csr+id_csr, upper=False, transpose=True).solution
  print("test6\n",res)
```
leading to
```
Default GPU is Tesla T4
Running on cpu
id_csr:
 tensor(crow_indices=tensor([0, 1, 2, 3, 4]),
       col_indices=tensor([0, 1, 2, 3]),
       values=tensor([1., 1., 1., 1.]), size=(4, 4), nnz=4,
       layout=torch.sparse_csr) 
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
A_dense:
 tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [2.6085, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.1871, 0.0000, 0.0000],
        [0.0000, 0.9922, 0.0000, 0.0000]])
b:
 tensor([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]])
test1
 tensor([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]])
test2
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test3
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test4
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test5
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test6
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
Running on cuda
id_csr:
 tensor(crow_indices=tensor([0, 1, 2, 3, 4]),
       col_indices=tensor([0, 1, 2, 3]),
       values=tensor([1., 1., 1., 1.]), device='cuda:0', size=(4, 4), nnz=4,
       layout=torch.sparse_csr) 
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]], device='cuda:0')
A_dense:
 tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [2.6085, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.1871, 0.0000, 0.0000],
        [0.0000, 0.9922, 0.0000, 0.0000]], device='cuda:0')
b:
 tensor([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]], device='cuda:0')
test1
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test2
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test3
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test4
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test5
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test6
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
```

As a side note, [torch.linalg.solve_triangular](https://pytorch.org/docs/stable/generated/torch.linalg.solve_triangular.html) does not provide a `transpose` flag anymore but does provide a `left` one. It might be good to provide an example there about using `left` in lieu of `transpose`.

### Versions

Google colab updated for PyTorch 1.13
```
Collecting environment information...
PyTorch version: 1.13.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.6
Libc version: glibc-2.26

Python version: 3.7.15 (default, Oct 12 2022, 19:14:55)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.10.133+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: True
CUDA runtime version: 11.2.152
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] torch==1.13.0+cu117
[pip3] torchaudio==0.13.0+cu117
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.14.0
[pip3] torchvision==0.14.0+cu117
[conda] Could not collect
```

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jianyuh @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

**Comment by nikitaved** [comment]
I cannot reproduce with:

```
PyTorch version: 1.14.0a0+gitaa914a0
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (conda-forge gcc 10.4.0-19) 10.4.0
Clang version: Could not collect
CMake version: version 3.24.2
Libc version: glibc-2.35

Python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:36:39) [GCC 10.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-48-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2060
GPU 1: NVIDIA GeForce RTX 2060

Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] mypy==0.960
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torch==1.14.0a0+git5078a72
[conda] magma-cuda117             2.6.1                         1    pytorch
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torch                     1.14.0a0+git5078a72           dev_0    <develop>

```

**Comment by mingfeima** [comment]
WIP.

**Comment by gajanan-choudhary** [comment]
This issue in TRSM in case of transpose + lower-triangular + unit diagonal treatment giving incorrect results has been reproduced on Intel oneMKL's side. I'm working on a fix for this.

**Comment by cpuhrsch** [comment]
@pearu - can you send a PR to disable this functionality until it's been fixed on Intel's side?

@gajanan-choudhary - do you have a timeline for this?

**Comment by nikitaved** [comment]
I will submit a temporary fix that materializes diagonal elements whenever `unitriangular=True`, up until the bug is fixed on the MKL's side.

**Comment by gajanan-choudhary** [comment]
> @pearu - can you send a PR to disable this functionality until it's been fixed on Intel's side?
> 
> @gajanan-choudhary - do you have a timeline for this?

The fix will be available in the next available public release version of oneMKL (2023.1), which is likely about a couple of months from now.

**Comment by ILoveSE** [comment]
@gajanan-choudhary  @nikitaved  Hello, is this fix already updated and available?
In the PyTorch 1.13.1 version, I use `torch.Tensor.triangular_solve` to conduct similar tests and found similar problems: using CPU calculations under `upper=False, transpose=True, unitrian=True` parameters will get wrong results.

<details>
  <summary>Error Output</summary>
  <pre><code>
  Running PyTorch version: 1.13.1
Default GPU is NVIDIA GeForce RTX 3090
Running on cpu
id_csr:
 tensor(crow_indices=tensor([0, 1, 2, 3, 4]),
       col_indices=tensor([0, 1, 2, 3]),
       values=tensor([1., 1., 1., 1.]), size=(4, 4), nnz=4,
       layout=torch.sparse_csr) 
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
A_dense:
 tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [2.6085, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.1871, 0.0000, 0.0000],
        [0.0000, 0.9922, 0.0000, 0.0000]])
b:
 tensor([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]])
test1
 tensor([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]])
test2
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test3
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test4
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test5
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
test6
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]])
Running on cuda
id_csr:
 tensor(crow_indices=tensor([0, 1, 2, 3, 4]),
       col_indices=tensor([0, 1, 2, 3]),
       values=tensor([1., 1., 1., 1.]), device='cuda:0', size=(4, 4), nnz=4,
       layout=torch.sparse_csr) 
 tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]], device='cuda:0')
A_dense:
 tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [2.6085, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.1871, 0.0000, 0.0000],
        [0.0000, 0.9922, 0.0000, 0.0000]], device='cuda:0')
b:
 tensor([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]], device='cuda:0')
test1
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test2
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test3
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test4
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test5
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')
test6
 tensor([[ 1.4677,  1.4677],
        [-0.1793, -0.1793],
        [ 1.0000,  1.0000],
        [ 1.0000,  1.0000]], device='cuda:0')

  </code></pre>
</details>

I want to ask: does this fix ( in oneMKL 2023.1 ) also work for similar bugs with `torch.Tensor.triangular_solve`?

Thank you very much!

```

import torch
print(f'Running PyTorch version: {torch.__version__}')

torchdevices = [torch.device('cpu')]
if torch.cuda.is_available():
  torchdevices.append( torch.device('cuda') )
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))

for torchdevice in torchdevices:
  print('Running on ' + str(torchdevice))

  A_dense = torch.tensor([[0.0000, 0.0000, 0.0000, 0.0000],
          [2.6085, 0.0000, 0.0000, 0.0000],
          [0.0000, 0.1871, 0.0000, 0.0000],
          [0.0000, 0.9922, 0.0000, 0.0000]], device=torchdevice)
  A_csr = A_dense.to_sparse_csr()

  id_csr = torch.sparse_csr_tensor(torch.arange(4+1), torch.arange(4), torch.ones(4, device=torchdevice), (4,4), device=torchdevice)
  print("id_csr:\n",id_csr,"\n",id_csr.to_dense())

  b = torch.ones(4, 2, device=torchdevice)

  print("A_dense:\n",A_dense)
  print("b:\n",b)

  res = torch.Tensor.triangular_solve(b, A_csr, upper=False, transpose=True, unitriangular=True).solution
  print("test1\n",res)

  res = torch.Tensor.triangular_solve(b, A_csr.t().to_sparse_csr(), upper=True, unitriangular=True).solution
  print("test2\n",res)

  res = torch.Tensor.triangular_solve(b, A_dense, upper=False, transpose=True, unitriangular=True).solution
  print("test3\n",res)

  res = torch.linalg.solve_triangular(A_dense.t(), b, upper=True, unitriangular=True)
  print("test4\n",res)

  res = torch.linalg.solve_triangular(A_dense, b.t(), upper=False, left=False, unitriangular=True).t()
  print("test5\n",res)

  res = torch.Tensor.triangular_solve(b, A_csr+id_csr, upper=False, transpose=True).solution
  print("test6\n",res)
```


<details>
  <summary>pytorch 1.13.1</summary>
  <pre><code>
  [pip3] numpy==1.22.3
  [pip3] torch==1.13.1
  [pip3] torchelastic==0.2.2
  [pip3] torchtext==0.14.1
  [pip3] torchvision==0.14.1
  [conda] blas                      1.0                         mkl  
  [conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
  [conda] mkl                       2021.4.0           h06a4308_640  
  [conda] mkl-service               2.4.0           py310h7f8727e_0  
  [conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
  [conda] mkl_random                1.2.2           py310h00e6091_0  
  [conda] numpy                     1.22.3          py310hfa59a62_0  
  [conda] numpy-base                1.22.3          py310h9585f30_0  
  [conda] pytorch                   1.13.1          py3.10_cuda11.6_cudnn8.3.2_0    pytorch
  [conda] pytorch-cuda              11.6                 h867d48c_1    pytorch
  [conda] pytorch-mutex             1.0                        cuda    pytorch
  [conda] torchelastic              0.2.2                    pypi_0    pypi
  [conda] torchtext                 0.14.1                    py310    pytorch
  [conda] torchvision               0.14.1              py310_cu116    pytorch</code></pre>
</details>

**Comment by nikitaved** [comment]
@ILoveSE , yes, it is fixed, albeit at a performance price in https://github.com/pytorch/pytorch/pull/93352. Take a look at the release candidate for PT2 and/or the latest nightly

**Comment by ILoveSE** [comment]
> @ILoveSE , yes, it is fixed, albeit at a performance price in #93352. Take a look at the release candidate for PT2 and/or the latest nightly

Thank you very much for your explanation! I'll keep an eye out for 2.0 releases.
